crypto/x509/x_attrib.c

```diff
@@ -56,3 +57,228 @@ X509_ATTRIBUTE *X509_ATTRIBUTE_create(int nid, int atrtype, void *value)
     ASN1_TYPE_free(val);
     return NULL;
 }
+
+static int print_hex(BIO *out, unsigned char *buf, int len)
+{
+    int i;
+
+    for (i = 0; i < len; i++) {
+        if (BIO_printf(out, "%02X ", buf[i]) <= 0) {
+            return 0;
+        }
+    }
+    return 1;
+}
+
+static int asn1_integer_print_bio(BIO *bio, const ASN1_INTEGER *num)
+{
+    BIGNUM *num_bn;
+    int result = 0;
+    char *hex;
+
+    num_bn = ASN1_INTEGER_to_BN(num, NULL);
+    if (num_bn == NULL)
+        return -1;
+    if ((hex = BN_bn2hex(num_bn)) != NULL) {
+        result = BIO_write(bio, "0x", 2) > 0;
+        result = result && BIO_write(bio, hex, strlen(hex)) > 0;
+        OPENSSL_free(hex);
+    } else {
+        return -1;
+    }
+    BN_free(num_bn);
+
+    return result;
+}
+
+static int print_oid (BIO *out, ASN1_OBJECT *oid) {
+    const char *ln;
+    char objbuf[80];
+    int rc;
+
+    if (OBJ_obj2txt(objbuf, sizeof(objbuf), oid, 1) <= 0)
+        return 0;
+    ln = OBJ_nid2ln(OBJ_obj2nid(oid));
+    rc = (ln != NULL)
+           ? BIO_printf(out, "%s (%s)", objbuf, ln)
+           : BIO_printf(out, "%s", objbuf);
+    if (rc < 0)
+        return 0;
+    return 1;
+}
+
+int ossl_print_attribute_value(BIO *out,
+                               int obj_nid,
+                               const ASN1_TYPE *av,
+                               int indent)
+{
+    ASN1_STRING *str;
+    unsigned char *value;
+    X509_NAME *xn = NULL;
+    int64_t int_val;
+
+    /*
+     * This switch-case is only for syntaxes that are not encoded as a single
+     * primitively-constructed value universal ASN.1 type.
+     */
+    switch (obj_nid) {
+    case NID_undef: /* Unrecognized OID. */
+        break;
+    /* Attribute types with DN syntax. */
+    case NID_member:
+    case NID_roleOccupant:
+    case NID_seeAlso:
+    case NID_manager:
+    case NID_documentAuthor:
+    case NID_secretary:
+    case NID_associatedName:
+    case NID_dITRedirect:
+    case NID_owner:
+        value = av->value.sequence->data;
+        xn = d2i_X509_NAME(NULL,
+                           (const unsigned char**)&(av->value.sequence->data),
```

Comment:
Why don't you send directly 'value' ptr here ?
It will spare you `av->value.sequence->data` pointer restoration a few lines below.